server/routerlicious/packages/services-ordering-rdkafka/src/rdkafkaBase.ts

```diff
@@ -91,15 +117,43 @@ export abstract class RdkafkaBase extends EventEmitter {
 		setTimeout(() => void this.initialize(), 1);
 	}
 
-	protected abstract connect(): void;
+	protected abstract connect(): Promise<void>;
+
+	protected async setOauthBearerTokenIfNeeded(client: kafkaTypes.Client<any>) {
+		if (!this.options.oauthBearerConfig?.tokenProvider) {
+			return;
+		}
+
+		Lumberjack.info("Setting up kafka client with oauthbearer token provider");
+
+		const tokenResponse = await this.options.oauthBearerConfig.tokenProvider();
+
+		if (!tokenResponse?.tokenStr || !tokenResponse?.refreshInMs) {
+			Lumberjack.error("Token provider returned malformed return value");
+			throw new Error("Token provider returned malformed return value");
+		}
+
+		if (tokenResponse.refreshInMs <= 0) {
+			Lumberjack.error("Token provider returned zero or negative 'refreshInMs'");
+			throw new Error("Token provider returned zero or negative 'refreshInMs'");
+		}
+
+		client.setOauthBearerToken(tokenResponse.tokenStr);
+
+		setTimeout(() => {
+			this.setOauthBearerTokenIfNeeded(client).catch((err) => {
```

Comment:
Should we use runWithRetry here? What's the impact of this failing?

Reply:
I wanted to keep it as simple as possible on the FF side, giving tokenProvider() implementors a decision on how they want to retry the token retrieval. If the tokenProvider() fails or setting of a token fails within node-rdkafka, it would mean the service completely stops due to old token expiring. Which means something is completely wrong and there's need for investigation on why tokenProvider() failed to provide a token or the setOauthBearerTokenIfNeeded() is broken and there has to be a fix for it. 

For FRS, I added retry logic in tokenProvider()'s implementation, and we can control it from that side.